.pre-commit-config.yaml

```diff
@@ -84,12 +84,22 @@ repos:
     -   id: pyright
         name: pyright
         entry: pyright
+        # note: assumes python env is setup and activated
         language: node
         pass_filenames: false
         types: [python]
         stages: [manual]
-        # note: keep version in sync with .github/workflows/code-checks.yml
         additional_dependencies: ['pyright@1.1.247']
+-   repo: local
+    hooks:
+    -   id: mypy
+        name: mypy
+        entry: mypy
+        # note: assumes python env is setup and activated
+        language: system
```

Comment:
would it work to use the mypy hook, and add `numpy` (and whatever else is needed for types) in `additional_dependencies`? I tried this a while ago and there some were inconsistencies with how mypy runs on the whole project versus as a hook, but if we were able to get it to pass as a hook, then it'd solve a whole bunch of typing headaches

I've seen a bunch of other projects do this, and it seems pretty reliable

Reply:
I would like that! I think the only downside is that the current pre-commit run is very fast as it doesn't need to install all the dependencies from environment.yml. I'm not familiar with pre-commit, is there a nice way of sharing the exact dependency versions (basically everything in environment.yml)?

I guess to get pyright working, we would need to use its `venv` argument to point it to pre-commit/mypy's environment (I don't think pre-commit allows installing python dependencies for a node program).

Reply:
does it need to be everything in `environment.yml`? I would've thought that only a few packages would be necessary, like `numpy` and maybe `pytest`

> is there a nice way of sharing the exact dependency versions (basically everything in environment.yml)?

I don't think so, other than writing our own script (like the one that checks flake8 versions match)

Reply:
> does it need to be everything in `environment.yml`?

Probably almost everything: imported libraries might have type annotations (and pyright explicitly checks that it can find all imported modules).

I think I will just see what a mess it will be :) then we can decide on whether it is worth it or not

Reply:
😄 no worries then, `system` is fine, as long as the stage is manual (it used to not be `system`, and then there were a tonne of false-positives, and people would tend to just turn off `pre-commit`)

Reply:
Okay, let's try `system` or this PR! Two questions:

1. Does github action create a new shell when executing pre-commit: the system hook doesn't find mypy from the activated conda env?
2. It would be great to have only pyright and mypy in `manual`. By default all hooks are in all stages. Could change `default_stages`, then we need to call pre-commit only once for mypy+pyright (might reduce some overhead). Do you think that would be helpful? If so, what value should `default_stages` have?

Reply:
It is actually not that bad listing all dependencies for mypy, see the latest commit.

The issue is to programmatically find the python env that pre-commit created for mypy so that pyright can use it as well.

While it is possible to move mypy and later move pyright, I would really like to make sure that pyright uses the same env as mypy, as pyright errors when it encounters any missing modules. It is easy to forget a module, which is then resolved as `Any` (can hide errors).

Reply:
I thought I found an ugly way that should work on linux but it seems that pyright (and probably mypy too?) needs pandas to be compiled for some of its checks, see the error in pre-commit now.

Compiling pandas in the pre-commit is a little bit too much (more dependencies and it takes much longer). Let's go back to the `system+manual` approach.

Reply:
@MarcoGorelli I'm mostly pattern matching when it comes to github action+pre-commit but I found a working `system+manual` version :)